dask_expr/expr.py

```diff
@@ -268,13 +303,45 @@ def simplify(self):
             else:
                 break
 
+        # Cache result
+        self._simplify_cache = {opt_group: expr}
         return expr
 
-    def _simplify_down(self):
-        return
+    def _simplify_down(self, opt_group: tuple(str)):
+        for opt in opt_group:
+            try:
+                out = getattr(self, f"_simplify_down_{opt}")()
+            except AttributeError:
+                out = None
+            if out is None:
+                continue
+            else:
+                return out
+        return self
 
-    def _simplify_up(self, parent):
-        return
+    def _simplify_up(self, parent: Expr, opt_group: tuple[str]):
+        for opt in opt_group:
+            try:
+                out = getattr(self, f"_simplify_up_{opt}")(parent)
+            except AttributeError:
+                out = None
+            if out is None:
+                continue
+            else:
+                return out
+        return parent
+
+    def _simplify_down_general(self):
+        return None
+
+    def _simplify_down_lower(self):
+        return None
+
+    def _simplify_up_general(self, parent: Expr):
+        return None
+
+    def _simplify_up_lower(self, parent: Expr):
+        return None
```

Comment:
We technically don't *need* these methods to exist, since `Expr._simplify_down`/`up` will return `None` when an `AttributeError` is encountered. However, the code is **much** faster when we don't need to handle missing-attribute errors often (which also means a dispatch-based approach would also be slower).

I still like the idea of handling the `AttributeError` case, because this allows users to add/use custom optimization types.